src/tap/cmd.rs

```diff
@@ -98,12 +99,27 @@ async fn run(
         );
     };
 
+    let start_time = Instant::now();
+    let stream_duration = match opts.duration {
+        Some(duration) => Duration::from_millis(duration),
+        None => Duration::MAX,
+    };
+
     // Loop over the returned results, printing out tap events.
     #[allow(clippy::print_stdout)]
     #[allow(clippy::print_stderr)]
     loop {
-        let message = stream.next().await;
-        if let Some(Some(res)) = message {
+        if start_time.elapsed() >= stream_duration {
+            return exitcode::OK;
+        }
+
+        let message = 
+            timeout(stream_duration - start_time.elapsed(), stream.next()).await;
+        if message.is_err() {
+            return exitcode::OK;
+        }
+
+        if let Ok(Some(Some(res))) = message {
```

Comment:
Given this pattern, we should probably use `match message` to encompass all three of the cases (including the `else` at the end of the function).